### PR TITLE
Enable .ptr .global .align attributes for kernel attributes for CUDA

### PR DESCRIPTION
**Pull request by LewisCrawford**
The current issue is PTX doesn't vectorise load and stores that can be vectorized.

We noticed that we were missing vectorization for sin, cos and power operations from LLVM resulting in lesser speedup. The reason is currently we don't generate any .ptr and .align attributes for kernel parameters in CUDA and the required alignment information is missing. This results in missing out on vectorization opportunities.
The change enables adding .align attribute for alignment information and .ptr attribute for kernel pointers in kernel parameters under the assumption that all kernel parameters pointers point to global memory space. This results in vectorization and boosting the speedup by ~2x.
.align is enabled only when the pointer has explicit alignment specifier.

PTX ISA doc - https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#kernel-parameter-attribute-ptr

This is a rework of the original patch proposed in #79646

### CODE REVIEW

**Comment by LewisCrawford** [comment]
Adding @jholewinski and @AlexMaclean as reviewers, as I've altered the original patch from #79646 quite a bit to address @Artem-B 's comments there to handle pointers explicitly annotated as shared/const more gracefully and add tests for them (even though CUDA assumes they do not occur).

In the original patch for CUDA, only explicitly aligned kernel ptr args got ".ptr .global .align N" added, and all others were ignored (no additional attrs, just emit a .u64 or u32 param).

In the version I've posted here, all pointers get ".ptr".the address space is set to .global/.shared/.const if specified explicitly (or set to .global if not specified on CUDA). If the pointer has an alignment explicitly set as anything other than 1, it will get a ".align" value added. Otherwise, on CUDA the pointer has no .align, but on CL, it gets .align 1.

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp
Comment (AlexMaclean) [inline llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp:1632]: If the alignment is explicitly set to 1 wouldn't we still want to emit the directive?

File: llvm/test/CodeGen/NVPTX/kernel-param-align.ll
Comment (AlexMaclean) [inline llvm/test/CodeGen/NVPTX/kernel-param-align.ll:18]: Can the bodies of these functions be completely removed?

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp
Comment (AlexMaclean) [inline llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp:1605]: use unsigned for addrSpace

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp
Comment (AlexMaclean) [inline llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp:1618]: This is likely just reflective of my own ignorance, but I'm still a bit confused as to why we should add ".global" to generic pointers in cuda?

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp
Comment (AlexMaclean) [inline llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp:1610]: nit: it looks like in some cases we may get double spaces here, can this be avoided?

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp
Comment (jholewinski) [reply llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp:1632]: If an explicit alignment is specified in the IR, it should be propagated to PTX. The default alignment in PTX is 4 bytes; we need to preserve 1- and 2-byte alignment to maintain correctness.

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp
Comment (jholewinski) [reply llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp:1618]: I'm inclined to say we should always honor the LLVM address space. If the front-end wants to assume a pointer is in the global space, it can use `addrspace(1)` on the function argument.

File: llvm/test/CodeGen/NVPTX/kernel-param-align.ll
Comment (LewisCrawford) [reply llvm/test/CodeGen/NVPTX/kernel-param-align.ll]: Done.

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp
Comment (LewisCrawford) [reply llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp]: I agree that makes more sense. I wasn't sure whether there was a good reason for avoiding emitting .align 1 in the original patch, so I kept that behaviour intact.

I've changed the code to emit all explicitly specified alignments on CUDA now.

Do we still need to preserve the CUDA vs OpenCL behavioural difference of emitting .align 1 for CL by default, and having no .align specifier for CUDA by default if no align is specified too, or can they both be changed to just not emit the alignment if it is not specified?

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp
Comment (LewisCrawford) [reply llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp]: The original patch which gave us perf improvements did so by assuming all pointers used as kernel args are .global. 

However, I have been struggling to reproduce that case locally myself, and as far as I can tell from the original discussions, it might only be the missing .align info that was needed anyway (as the address space can be inferred from things like `cvta.to.global.u64 ` and `ld.global.nc` already).

I agree that it is much cleaner to just honour the LLVM address space, rather than making undocumented and potentially invalid assumptions. If there turn out to be strong perf reasons for assuming .global for generic pointers later on, we should probably document this implicit assumption better somewhere.

I've now also added support for emitting .local here, as the PTX spec says this is technically allowed here too: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#kernel-parameter-attribute-ptr

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp
Comment (LewisCrawford) [reply llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp]: Fixed.

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp
Comment (LewisCrawford) [reply llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp]: I've moved the PTy->getAddressSpace() directily into the switch() now.

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp]: +1 for emitting alignment when it's known. If it turns out to be not optimal, that would be a separate issue to fix.

Comment (LewisCrawford) [reply llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp]: The latest version is:
`          const bool IsCUDA =
              static_cast<NVPTXTargetMachine &>(TM).getDrvInterface() ==
              NVPTX::CUDA;

          MaybeAlign ParamAlign = I->getParamAlign();
          if (ParamAlign.has_value() || !IsCUDA)
            O << ".align " << ParamAlign.valueOrOne().value() << " ";`

So, it always emits the alignment when known. 
When unknown, OpenCL defaults to 1, and CUDA defaults to "do not emit align info" (implicitly assume 4 according to the PTX spec: https://docs.nvidia.com/cuda/parallel-thread-execution/#kernel-function-parameter-attributes ).

Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp]: I'm a little unsure about why it is safe to implicitly assume 4 for CUDA. What are the semantics for LLVM IR when no alignment is specified? Shouldn't we faithfully translate these into PTX?

**Review by AlexMaclean — APPROVED** [review]
With the latest changes, this looks good to me. I'm not familiar enough with the semantics of llvm IR to be sure that the default alignment is 1, but that seems like a reasonable and conservative way to go. Please wait for an additional approval before landing.

**Review by Artem-B — APPROVED** [review]

**Comment by tom91136** [comment]
Hello, 
I've just tested this patch and I'm getting `CUDA_ERROR_INVALID_IMAGE` for any ptr kernel param attributes that isn't `.global` or empty.
I've reduced it to the  following PTX which fails on load and reports `CUDA_ERROR_INVALID_IMAGE`

```cpp
#include <cstdio>
#include <cuda.h>
#define CUDA_CHECK(call)                                                                                                                   \
  do {                                                                                                                                     \
    if (const CUresult res = call; res != CUDA_SUCCESS) {                                                                                  \
      const char *err;                                                                                                                     \
      cuGetErrorName(res, &err);                                                                                                           \
      std::fprintf(stderr, "Error %s:%d: %s\n", __FILE__, __LINE__, err);                                                                  \
      std::exit(res);                                                                                                                      \
    }                                                                                                                                      \
  } while (0)
int main() {
  CUDA_CHECK(cuInit(0));
  CUdevice dev;
  CUDA_CHECK(cuDeviceGet(&dev, 0));
  CUcontext ctx;
  CUDA_CHECK(cuCtxCreate(&ctx, 0, dev));
  CUmodule m;
  CUfunction f;
  const char *ptx = R"PTX(
.version 7.8
.target sm_89
.address_size 64
.visible .entry kernel(
  .param .u64 .ptr         .align 1 kernel_param_0,
  .param .u64 .ptr .global .align 1 kernel_param_2,
  .param .u64 .ptr .shared .align 1 kernel_param_3
){
  ret;
}
)PTX";
  CUDA_CHECK(cuModuleLoadData(&m, ptx));
  CUDA_CHECK(cuModuleGetFunction(&f, m, "kernel"));
  return 0;
}

```
If we take out the `.shared`, then the module loads. 
Reading the PTX spec, I think attributes like `.shared` should be allowed so not sure why it's failing.
The behaviour before this patch doesn't append address space attributes so it worked.

**Comment by Artem-B** [comment]
I can't reproduce it. AFAICT, `.shared` works fine with ptxas as far back as cuda-11: https://godbolt.org/z/W1hMxWTKd

**Comment by tom91136** [comment]
ptxas also worked on my end. Can you try the reproducer and see if it loads the module?

**Comment by LewisCrawford** [comment]
Here's a patch to add the extra error checking: https://github.com/llvm/llvm-project/pull/138706

**Pull request merged**
